retrofit/src/main/java/retrofit/OkHttpCall.java

```diff
@@ -109,14 +120,93 @@ private void callSuccess(Response<T> response) {
     }
     this.rawCall = rawCall;
 
-    return parseResponse(rawCall.execute());
+    final long start = System.nanoTime();
+    com.squareup.okhttp.Response response = rawCall.execute();
+    long elapsedTime = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start);
+
+    return parseResponse(response, elapsedTime);
   }
 
   private com.squareup.okhttp.Call createRawCall() {
-    return client.newCall(requestFactory.create(args));
+    Request request = requestFactory.create(args);
+    if (logger.level().log()) {
+      request = logAndReplaceRequest(request, args);
+    }
+    return client.newCall(request);
   }
 
-  private Response<T> parseResponse(com.squareup.okhttp.Response rawResponse) throws IOException {
+  private Request logAndReplaceRequest(Request request, Object[] args)  {
+    logger.log(String.format("---> HTTP %s %s", request.method(), request.urlString()));
+
+    if (logger.level().ordinal() >= LogLevel.HEADERS.ordinal()) {
+      if (request.headers().size() != 0) {
+        logger.log(request.headers().toString());
+      }
+
+      String bodySize = "no";
+      RequestBody body = request.body();
+      if (body != null) {
+        final MediaType contentType = body.contentType();
+        if (contentType != null) {
+          logger.log("Content-Type: " + contentType);
+        }
+
+        long contentLength = -1;
+        try {
+          contentLength = body.contentLength();
+        } catch (IOException e) {
+          e.printStackTrace();
+        }
+        bodySize = contentLength + "-byte";
+        if (contentLength != -1) {
+          logger.log("Content-Length: " + contentLength);
+        }
+
+        if (logger.level().ordinal() >= LogLevel.FULL.ordinal()) {
+          if (request.headers().size() != 0) {
+            logger.log("");
+          }
+          try {
+            final Buffer buffer = new Buffer();
+            body.writeTo(buffer);
+            logger.log(buffer.snapshot().utf8());
+            request = request.newBuilder()
+                .method(request.method(), new RequestBody() {
+                  @Override public MediaType contentType() {
+                    return contentType;
+                  }
+
+                  @Override public void writeTo(BufferedSink sink) throws IOException {
+                    sink.write(buffer, buffer.size());
+                  }
+                })
+                .build();
+          } catch (IOException e) {
+            e.printStackTrace();
```

Comment:
Looks suspicious, if we can not read request's body due to `IOException` (bad connection, etc), this exception should be thrown up to the caller, shouldn't it?